File: quantum-resistant/symmetricPrimitives.js
// Quantum-resistant symmetric primitives: AES-256-GCM, SHA-512, HMAC-SHA-256.
'use strict';
const crypto = require('crypto');

function aes256Gcm(plaintext) {
  const key = crypto.randomBytes(32);
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  const ct = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  return Buffer.concat([iv, cipher.getAuthTag(), ct]);
}

const sha512 = (data) => crypto.createHash('sha512').update(data).digest('hex');
const hmacSha256 = (key, data) => crypto.createHmac('sha256', key).update(data).digest('hex');

console.log('AES-256-GCM output len:', aes256Gcm(Buffer.from('data')).length);
console.log('SHA-512:', sha512('data').slice(0, 32));
console.log('HMAC-SHA-256:', hmacSha256('key', 'data').slice(0, 32));
module.exports = { aes256Gcm, sha512, hmacSha256 };

File: quantum-resistant/symmetric_primitives.rs
// Quantum-resistant symmetric primitives: AES-256-GCM, SHA-512, HMAC-SHA-256.
// crates: aes-gcm, sha2, hmac
use aes_gcm::aead::{Aead, AeadCore, KeyInit, OsRng};
use aes_gcm::Aes256Gcm;
use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256, Sha512};

pub fn aes256_gcm(plaintext: &[u8]) -> Vec<u8> {
    let key = Aes256Gcm::generate_key(&mut OsRng);
    let cipher = Aes256Gcm::new(&key);
    let nonce = Aes256Gcm::generate_nonce(&mut OsRng);
    let mut out = nonce.to_vec();
    out.extend_from_slice(&cipher.encrypt(&nonce, plaintext).expect("encrypt"));
    out
}

pub fn sha512(data: &[u8]) -> [u8; 64] {
    let mut hasher = Sha512::new();
    hasher.update(data);
    hasher.finalize().into()
}

pub fn hmac_sha256(key: &[u8], data: &[u8]) -> Vec<u8> {
    let mut mac = <Hmac<Sha256>>::new_from_slice(key).expect("any key length");
    mac.update(data);
    mac.finalize().into_bytes().to_vec()
}
